File: test-suite/src/test/java/io/micronaut/security/docs/sensitiveendpointrule/SensitiveEndpointRuleReplacementTest.java
package io.micronaut.security.docs.sensitiveendpointrule;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.security.testutils.authprovider.MockAuthenticationProvider;
import io.micronaut.security.testutils.authprovider.SuccessAuthenticationScenario;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import java.util.Collections;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertEquals;

@Property(name = "micronaut.security.oauth2.enabled", value = StringUtils.FALSE)
@Property(name = "endpoints.beans.enabled", value = StringUtils.TRUE)
@Property(name = "endpoints.beans.sensitive", value = StringUtils.TRUE)
@Property(name = "spec.name", value = "SensitiveEndpointRuleReplacementTest")
@MicronautTest
public class SensitiveEndpointRuleReplacementTest {

    @Inject
    @Client("/")
    HttpClient httpClient;

    @Test
    void testAccessingASensitiveEndpointWithAuthenticationAndASensitiveEndpointRuleReplacementWorks() {
        BlockingHttpClient client = httpClient.toBlocking();
        Executable e = () -> client.exchange(HttpRequest.GET("/beans"));
        HttpClientResponseException thrown = assertThrows(HttpClientResponseException.class, e);
        assertEquals(HttpStatus.UNAUTHORIZED, thrown.getStatus());
        e = () -> client.exchange(HttpRequest.GET("/beans").basicAuth("user", "password"));
        assertDoesNotThrow(e);
    }

    @Singleton
    @Requires(property = "spec.name", value = "SensitiveEndpointRuleReplacementTest")
    static class AuthenticationProviderUserPassword extends MockAuthenticationProvider {
        AuthenticationProviderUserPassword() {
            super(Collections.singletonList(new SuccessAuthenticationScenario("user")));
        }
    }
}

File: test-suite-python/src/test/python/micronaut/security/docs/sensitiveendpointrule/SensitiveEndpointRuleReplacementTest.py
from typing import Annotated

from jakarta.inject import Inject, Singleton
from micronaut.context.annotation import Property, Requires
from micronaut.core.util import StringUtils
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.security.authentication import AuthenticationFailureReason, AuthenticationRequest, AuthenticationResponse
from micronaut.security.authentication.provider import HttpRequestAuthenticationProvider
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test


@Property(name="micronaut.security.oauth2.enabled", value=StringUtils.FALSE)
@Property(name="endpoints.beans.enabled", value=StringUtils.TRUE)
@Property(name="endpoints.beans.sensitive", value=StringUtils.TRUE)
@Property(name="spec.name", value="SensitiveEndpointRuleReplacementTest")
@MicronautTest
class SensitiveEndpointRuleReplacementTest:
    httpClient: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_accessing_a_sensitive_endpoint_with_authentication_and_a_sensitive_endpoint_rule_replacement_works(self):
        client = self.httpClient.toBlocking()
        try:
            client.exchange(HttpRequest.GET("/beans"))
            assert False, "HttpClientResponseException expected"
        except HttpClientResponseException as e:
            assert e.getStatus() == HttpStatus.UNAUTHORIZED
        client.exchange(HttpRequest.GET("/beans").basicAuth("user", "password"))


@Singleton
@Requires(property="spec.name", value="SensitiveEndpointRuleReplacementTest")
class AuthenticationProviderUserPassword(HttpRequestAuthenticationProvider):

    def authenticate(self, requestContext: HttpRequest, authRequest: AuthenticationRequest) -> AuthenticationResponse:
        if authRequest.getIdentity() == "user":
            return AuthenticationResponse.success("user")
        return AuthenticationResponse.failure(AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH)
